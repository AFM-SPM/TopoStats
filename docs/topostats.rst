topostats package
=================

Subpackages
-----------

.. toctree::
   :maxdepth: 4

   topostats.logs
   topostats.tracing

Submodules
----------

.. toctree::
   :maxdepth: 4

   topostats.entry_point
   topostats.filters
   topostats.grains
   topostats.grainstats
   topostats.io
   topostats.plotting
   topostats.plottingfuncs
   topostats.processing
   topostats.run_topostats
   topostats.scars
   topostats.statistics
   topostats.theme
   topostats.thresholds
   topostats.utils
   topostats.validation

Module contents
---------------

.. automodule:: topostats
   :members:
   :undoc-members:
   :show-inheritance:
