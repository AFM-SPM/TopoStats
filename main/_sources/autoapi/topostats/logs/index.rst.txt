topostats.logs
==============

.. py:module:: topostats.logs

.. autoapi-nested-parse::

   
   Initialise login module.
















   ..
       !! processed by numpydoc !!


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/topostats/logs/logs/index


