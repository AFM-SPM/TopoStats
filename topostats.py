"""
File for tracking hard-coded variables (for now)
"""

import argparse

### List of hard-coded variables in pywgwytracing.py

# global variables
gwy_colobar_ln48 = "/module/pixmap/ztype"
gwy_linematchmethod_ln51= '/module/linematch/method'
gwy_linematchmaxdegree_ln52 = "/module/linematch/max_degree"
gwy_polylevel_coldegree_ln53 = "/module/polylevel/col_degree"

# heightediting()
datafield_grains_mark_height_ln128 = int()
gwy_polylevel_masking_ln131 = int()
gwy_linematch_masking_ln135 = int()
filterwidth_ln144 = float()

# editfile()
datafield_markheight_ln170 = int()
gwy_polylevel_masking_ln173 = int()
gwy_linematch_masking_ln177 = gwy_linematch_masking_ln135
gwy_polylevel_masking_ln181 = int()
datafield_gaussfilter_ln195 = int()

# grainfinding()
grain_gauss_ln211 = float()
datafield_mask_outliers_ln220 = float()

# remove_large_objects()
datafield_mask_outliers_ln257 = float()

# remove_small_objects()
datafield_mask_outliers_ln283 = float()

# grain_thinning()
gaussian_size_ln488 = float()

# savefiles()
savefile_type_ln535 = ["/module/pixmap/xytype"]

# saveunknownfiles()
savefile_type_ln573 = ["/module/pixmap/xytype"]

# savecroppedfiles()
savefile_type_ln645 = ["/module/pixmap/xytype"]

# __main__()
path_ln709 = str()
fileend_ln712 = []
filetype_ln713 = str()
extension_ln715 = str()
minheightscale_ln717 = float()
maxheightscale_ln718 = float()
minarea_ln722 = float()
maxdeviation_ln726 = float()
mindeviation_ln727 = float()
cropwidth_ln733 = float()
splitwidth_ln734 = float()
bins_ln736 = int()
channels_ln755 = list()


### dnatracing.py

# getNumpyArraysfromGwyddion()
sigma_ln94 = float()

# getDisorderedTrace()
sigma_ln109 = float()

# purgeObviousCrap()
maxlen_ln126 = int()

# getFittedTraces()
index_width_ln212 = int()

# getSplinedTraces()
step_size_ln376 = float()
interp_size_ln377 = float()

### tracingfuncs.py

# I didn't find any hard-coded variables so far

### statsplotting.py
### TFOplot.py
### traceplotting.py

# not checking these for now due to ongoing merging project
