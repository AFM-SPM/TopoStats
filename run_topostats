#!/usr/bin/env python3
"""Topotracing"""
import argparse as arg
from asyncio.log import logger
from collections import defaultdict

# from email.policy import default
from functools import partial
from multiprocessing import Pool
from pathlib import Path
from typing import Union, Dict

import pandas as pd
import numpy as np
from tqdm import tqdm
import numpy as np

from topostats.filters import Filters
from topostats.grains import Grains
from topostats.grainstats import GrainStats, get_grainstats
from topostats.io import read_yaml
from topostats.logs.logs import setup_logger, LOGGER_NAME
from topostats.plottingfuncs import plot_and_save
from topostats.thresholds import threshold
from topostats.tracing.dnatracing import dnaTrace
from topostats.utils import convert_path, find_images, update_config

LOGGER = setup_logger(LOGGER_NAME)


def create_parser() -> arg.ArgumentParser:
    """Create a parser for reading options."""
    parser = arg.ArgumentParser(
        description="Process AFM images. Additional arguments over-ride those in the configuration file."
    )
    parser.add_argument(
        "-c",
        "--config_file",
        dest="config_file",
        required=True,
        help="Path to a YAML configuration file.",
    )
    parser.add_argument(
        "-b",
        "--base_dir",
        dest="base_dir",
        type=str,
        required=False,
        help="Base directory to scan for images.",
    )
    parser.add_argument(
        "-f",
        "--file_ext",
        dest="file_ext",
        type=str,
        required=False,
        help="File extension to scan for.",
    )
    parser.add_argument(
        "--channel",
        dest="channel",
        type=str,
        required=False,
        help="Channel to extract.",
    )
    parser.add_argument(
        "-o",
        "--output_dir",
        dest="output_dir",
        type=str,
        required=False,
        help="Output directory to write results to.",
    )
    parser.add_argument(
        "-s",
        "--save_plots",
        dest="save_plots",
        type=bool,
        required=False,
        help="Whether to save plots.",
    )
    parser.add_argument(
        "-a",
        "--amplify_level",
        dest="amplify_level",
        type=float,
        required=False,
        help="Amplify signals by the given factor.",
    )
    parser.add_argument(
        "-t",
        "--threshold_method",
        dest="threshold_method",
        required=False,
        help="Method used for thresholding.",
    )
    parser.add_argument(
        "--otsu_threshold_multiplier",
        dest="otsu_threshold_multiplier",
        required=False,
        help="Factor to scale threshold during grain finding.",
    )
    parser.add_argument("-m", "--mask", dest="mask", type=bool, required=False, help="Mask the image.")
    parser.add_argument(
        "-q",
        "--quiet",
        dest="quiet",
        type=bool,
        required=False,
        help="Toggle verbosity.",
    )
    parser.add_argument(
        "-w",
        "--warnings",
        dest="quiet",
        type=str,
        required=False,
        help="Whether to ignore warnings.",
    )
    return parser


def process_scan(
    image_path: Union[str, Path] = None,
    channel: str = "Height",
    amplify_level: float = 1.0,
    filter_threshold_method: str = "otsu",
    filter_otsu_threshold_multiplier: Union[int, float] = 1.7,
    filter_threshold_std_dev=1.0,
    filter_threshold_abs_lower=None,
    filter_threshold_abs_upper=None,
    gaussian_size: Union[int, float] = 2,
    gaussian_mode: str = "nearest",
    absolute_smallest_grain_size=None,
    background: float = 0.0,
    grains_threshold_method: str = "otsu",
    grains_otsu_threshold_multiplier: Union[int, float] = 1.7,
    grains_threshold_std_dev=1.0,
    grains_threshold_abs_lower=None,
    grains_threshold_abs_upper=None,
    save_plots: bool = True,
    output_dir: Union[str, Path] = "output",
) -> None:
    """Process a single image, filtering, finding grains and calculating their statistics.

    Parameters
    ----------
    image_path : Union[str, Path]
        Path to image to process.
    channel : str
        Channel to extract and process, default 'height'.
    amplify_level : float
        Level to amplify image prior to processing by.
    threshold_method: str
        Method for determining threshold to mask values, default is 'otsu'.
    gaussian_size : Union[int, float]
        Minimum grain size in nanometers (nm).
    mode : str
        Mode for filtering (default is 'nearest').
    otsu_threshold_multiplier : Union[int, float]
        Factor by which lower threshold is to be scaled prior to masking.
    background : float
    save_plots : bool
        Flag as to whether to save plots to PNG files.
    output_dir : Union[str, Path]
        Path to output directory for saving results.

    Examples
    --------

    from topostats.topotracing import process_scan

    process_scan(image_path='minicircle.spm',
                 save_plots=True,
                 output_dir='output/')

    """
    LOGGER.info(f"Processing : {image_path}")
    # Filter Image :
    #
    # The Filters class has a convenience method that runs the instantiated class in full.
    filtered_image = Filters(
        image_path,
        threshold_method=filter_threshold_method,
        threshold_std_dev=filter_threshold_std_dev,
        threshold_absolute_lower=filter_threshold_abs_lower,
        threshold_absolute_upper=filter_threshold_abs_upper,
        channel=channel,
        amplify_level=amplify_level,
        output_dir=output_dir / image_path.stem / "filters",
    )
    filtered_image.filter_image()

    # Find Grains :
    #
    # The Grains class also has a convenience method that runs the instantiated class in full.
    grains = Grains(
        image=filtered_image.images["zero_averaged_background"],
        filename=filtered_image.filename,
        pixel_to_nm_scaling=filtered_image.pixel_to_nm_scaling,
        gaussian_size=gaussian_size,
        gaussian_mode=gaussian_mode,
        threshold_method=grains_threshold_method,
        otsu_threshold_multiplier=grains_otsu_threshold_multiplier,
        threshold_std_dev=grains_threshold_std_dev,
        threshold_absolute_lower=grains_threshold_abs_lower,
        threshold_absolute_upper=grains_threshold_abs_upper,
        absolute_smallest_grain_size=absolute_smallest_grain_size,
        background=background,
        base_output_dir=output_dir / filtered_image.filename / "grains",
    )
    grains.find_grains()

    # Grainstats :
    #
    # Ditto for the GrainStats class.
    grainstats = GrainStats(
        data=grains.images["gaussian_filtered"],
        labelled_data=grains.images["labelled_regions"],
        pixel_to_nanometre_scaling=filtered_image.pixel_to_nm_scaling,
        img_name=filtered_image.filename,
        output_dir=output_dir,
    )
    grain_statistics = grainstats.calculate_stats()

    # Run dnatracing
    # np.savetxt(Path("/home/neil/tmp/TopoStats/TOPOSTATS_PIXELS.csv"), filtered_image.images["pixels"], delimiter=",")
    # np.savetxt(Path("/home/neil/tmp/TopoStats/TOPOSTATS_GRAINS.csv"), grains.images["labelled_regions"], delimiter=",")
    # np.savetxt(
    #     Path("/home/neil/tmp/TopoStats/TOPOSTATS_GAUSSIAN_FILTERED.csv"),
    #     grains.images["gaussian_filtered"],
    #     delimiter=",",
    # )
    print(f"FROM run_topostats : {grains.images['labelled_regions']}")
    #    print(f"FROM run_topostats : {grains.images['labelled_regions'].tolist()}")
    dna_traces = dnaTrace(
        full_image_data=grains.images["gaussian_filtered"].T,
        grains=grains.images["labelled_regions"],
        afm_image_name=filtered_image.filename,
        pixel_size=filtered_image.pixel_to_nm_scaling,
        number_of_columns=grains.images["labelled_regions"].shape[0],
        number_of_rows=grains.images["labelled_regions"].shape[1],
    )

    # Optionally plot all stages
    if save_plots:
        # Filtering stage
        for plot_name, array in filtered_image.images.items():
            if plot_name not in ["scan_raw", "extracted_channel"]:
                PLOT_DICT[plot_name]["output_dir"] = Path(output_dir) / filtered_image.filename
                plot_and_save(array, **PLOT_DICT[plot_name])
        # Grain stage - only if we have grains
        if len(grains.region_properties) > 0:
            for plot_name, array in grains.images.items():
                PLOT_DICT[plot_name]["output_dir"] = Path(output_dir) / filtered_image.filename
                plot_and_save(array, **PLOT_DICT[plot_name])
            # Make a plot of coloured regions with bounding boxes
            plot_and_save(
                grains.images["coloured_regions"],
                Path(output_dir) / filtered_image.filename,
                **PLOT_DICT["bounding_boxes"],
                region_properties=grains.region_properties,
            )
            plot_and_save(
                grains.images["labelled_regions"],
                Path(output_dir) / filtered_image.filename,
                **PLOT_DICT["coloured_boxes"],
                region_properties=grains.region_properties,
            )
        # Grainstats
        # FIXME : Include this


def main():
    """Run processing."""

    # Parse command line options, load config and update with command line options
    parser = create_parser()
    args = parser.parse_args()
    config = read_yaml(args.config_file)
    config = update_config(config, args)

    LOGGER.info(f"Configuration file loaded from    : {args.config_file}")
    LOGGER.info(f'Scanning for images in            : {config["base_dir"]}')
    LOGGER.info(f'Output directory                  : {config["output_dir"]}')
    LOGGER.info(f'Looking for images with extension : {config["file_ext"]}')
    img_files = find_images(config["base_dir"])
    LOGGER.info(f'Images with extension {config["file_ext"]} in {config["base_dir"]} : {len(img_files)}')

    if config["quiet"]:
        LOGGER.setLevel("ERROR")
    # For debugging (as Pool makes it hard to find things when they go wrong)
    # for x in img_files:
    #     process_scan(
    #         image_path=x,
    #         channel=config["channel"],
    #         amplify_level=config["amplify_level"],
    #         threshold_method=config["threshold_method"],
    #         otsu_threshold_multiplier=config["grains"]["otsu_threshold_multiplier"],
    #         gaussian_size=config["grains"]["gaussian_size"],
    #         gaussian_mode=config["grains"]["gaussian_mode"],
    #         background=config["grains"]["background"],
    #         output_dir=config["output_dir"],
    #     )
    processing_function = partial(
        process_scan,
        channel=config["channel"],
        amplify_level=config["amplify_level"],
        filter_threshold_method=config["filter"]["threshold"]["method"],
        filter_otsu_threshold_multiplier=config["filter"]["threshold"]["otsu_multiplier"],
        filter_threshold_std_dev=config["filter"]["threshold"]["std_dev"],
        filter_threshold_abs_lower=config["filter"]["threshold"]["absolute"][0],
        filter_threshold_abs_upper=config["filter"]["threshold"]["absolute"][1],
        gaussian_size=config["grains"]["gaussian_size"],
        gaussian_mode=config["grains"]["gaussian_mode"],
        absolute_smallest_grain_size=config["grains"]["absolute_smallest_grain_size"],
        background=config["grains"]["background"],
        output_dir=config["output_dir"],
        grains_threshold_method=config["grains"]["threshold"]["method"],
        grains_otsu_threshold_multiplier=config["grains"]["threshold"]["otsu_multiplier"],
        grains_threshold_std_dev=config["grains"]["threshold"]["std_dev"],
        grains_threshold_abs_lower=config["grains"]["threshold"]["absolute"][0],
        grains_threshold_abs_upper=config["grains"]["threshold"]["absolute"][1],
    )

    with Pool(processes=config["cores"]) as pool:
        with tqdm(
            total=len(img_files),
            desc=f'Processing images from {config["base_dir"]}, results are under {config["output_dir"]}',
        ) as pbar:
            for _ in pool.imap_unordered(processing_function, img_files):
                pbar.update()


if __name__ == "__main__":
    main()
