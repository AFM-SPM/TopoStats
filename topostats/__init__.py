"""Topostats"""
from importlib.metadata import version

from .logs.logs import setup_logger

LOGGER = setup_logger()

release = version("topostats")
__version__ = ".".join(release.split("."[:2]))
