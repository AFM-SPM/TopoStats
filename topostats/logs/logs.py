"""
Standardise logging.
"""
import sys
from pathlib import Path
from datetime import datetime
import logging

# pylint: disable=assignment-from-no-return

start = datetime.now()
LOG_INFO_FORMATTER = logging.Formatter(
    fmt="[%(asctime)s] [%(levelname)-8s] [%(name)s] %(message)s", datefmt="%a, %d %b %Y %H:%M:%S"
)
LOG_ERROR_FORMATTER = logging.Formatter(
    fmt="[%(asctime)s] [%(levelname)-8s] [%(name)s] [%(filename)s] [%(lineno)s] %(message)s",
    datefmt="%a, %d %b %Y %H:%M:%S",
)

LOGGER_NAME = "topostats"


def setup_logger(log_name: str = LOGGER_NAME) -> logging.Logger:
    """Setup a standard logger.

    The logger for the module is initialised when the module is loaded (as this functions is called from
    __init__.py). This creates two stream handlers, one for general output and one for errors which are formatted
    differently (there is greater information in the error formatter). To use in modules import the 'LOGGER_NAME' and
    create a logger as shown in the Examples, it will inherit the formatting and direction of messages to the correct
    stream.

    Parameters
    ----------
    log_name : str
        Name under which logging information occurs.

    Returns
    -------
    logging.Logger
        Logger object.

    Examples
    --------

    To use the logger in (sub-)modules have the following.

        import logging
        from topostats.logs.logs import LOGGER_NAME

        LOGGER = logging.getLogger(LOGGER_NAME)

        LOGGER.info('This is a log message.')
    """
    out_stream_handler = logging.StreamHandler(sys.stdout)
    out_stream_handler.setLevel(logging.DEBUG)
    out_stream_handler.setFormatter(LOG_INFO_FORMATTER)

    err_stream_handler = logging.StreamHandler(sys.stderr)
    err_stream_handler.setLevel(logging.ERROR)
    err_stream_handler.setFormatter(LOG_ERROR_FORMATTER)

    file_handler = logging.FileHandler(Path().cwd().stem + f"-{start.strftime('%Y-%m-%d-%H-%M-%S')}.log")
    file_handler.setFormatter(LOG_ERROR_FORMATTER)

    logger = logging.getLogger(log_name)
    logger.setLevel(logging.INFO)
    logger.propagate = True
    if not logger.handlers:
        logger.addHandler(out_stream_handler)
        logger.addHandler(err_stream_handler)
        logger.addHandler(file_handler)

    return logger
