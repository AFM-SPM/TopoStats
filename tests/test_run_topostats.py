"""Test end-to-end running of topostats."""
import logging
from pathlib import Path

import pytest

from topostats.logs.logs import LOGGER_NAME
from topostats.run_topostats import main as run_topostats_main

BASE_DIR = Path.cwd()


@pytest.mark.parametrize("option", ("-h", "--help"))
def test_run_topostats_main_help(capsys, option) -> None:
    """Test the -h/--help flag to run_topostats."""
    try:
        run_topostats_main([option])
    except SystemExit:
        pass
    assert "Process AFM images." in capsys.readouterr().out


def test_run_topostats_process_all(caplog) -> None:
    """Test run_topostats completes without error when no arguments are given."""
    caplog.set_level(logging.INFO)
    # Explicitly force loading of topostats/default_config.yaml as I couldn't work out how to invoke process_all()
    # without any arguments as it defaults to 'sys.argv' as this is wrapped within pytest it picks up the arguments
    # pytest was invoked with (see thread on StackOverflow at https://stackoverflow.com/a/55260580/1444043)
    run_topostats_main(args=["--config", f"{BASE_DIR / 'topostats' / 'default_config.yaml'}"])
    assert "~~~~~~~~~~~~~~~~~~~~ COMPLETE ~~~~~~~~~~~~~~~~~~~~" in caplog.text
    assert "Successfully Processed^1    : 1 (100.0%)" in caplog.text


def test_run_topostats_process_debug(caplog) -> None:
    """Test run_topostats with debugging and check DEBUG messages are logged"""
    # Set the logging level of the topostats logger
    with caplog.at_level(logging.DEBUG, logger=LOGGER_NAME):
        run_topostats_main(args=["--config", f"{BASE_DIR / 'topostats' / 'default_config.yaml'}", "-l", "debug"])
        assert "Configuration after update         :" in caplog.text
        assert "File extension : .spm" in caplog.text
        assert "Images processed : 1" in caplog.text
        assert "~~~~~~~~~~~~~~~~~~~~ COMPLETE ~~~~~~~~~~~~~~~~~~~~" in caplog.text
