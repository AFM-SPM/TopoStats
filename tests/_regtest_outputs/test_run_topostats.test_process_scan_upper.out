                   centre_x    centre_y  radius_min  radius_max  radius_mean  radius_median  height_min  height_max  height_median  height_mean       volume        area  area_cartesian_bbox  smallest_bounding_width  smallest_bounding_length  smallest_bounding_area  aspect_ratio threshold  max_feret  min_feret  Contour Lengths  Circular  End to End Distance      Image Name
Molecule Number                                                                                                                                                                                                                                                                                                                                                                       
0                 18.707007   96.557844    5.704484   20.868161    15.487836      16.147027    0.806496    3.871971       2.068617     1.978976  1209.299322  611.073392          1137.465287                23.833883                 41.299524              984.327993      1.732807     upper  41.334203  23.833883        92.386012      True             0.000000  minicircle.spm
1                400.792704   92.681699    1.152017   21.957089    13.681248      14.459307    0.807036    3.185268       2.153416     2.015202  1252.583477  621.567064          1166.261877                27.021623                 39.031234             1054.687294      1.444444     upper  39.063733  27.021623        97.805725     False            15.753970  minicircle.spm
2                217.783200  173.222730    8.327740   28.522381    17.743418      16.365654    0.808289    3.088261       2.003851     1.895803  1138.118210  600.335681          1577.955492                24.191610                 54.065589             1307.933633      2.234890     upper  54.065681  24.191610       134.804884     False             0.698626  minicircle.spm
3                353.728516  207.406570    3.923541   25.897458    15.301596      15.862244    0.808945    3.103586       2.138866     2.016413  1214.461497  602.287992          1619.442104                17.774995                 49.449371              878.962307      2.781963     upper  49.462005  17.774995        91.466549     False            26.349396  minicircle.spm
4                300.802327  212.096860   10.146734   25.314812    17.927247      18.386595    0.807940    3.003616       1.926633     1.866040  1306.503578  700.147588          1719.742090                32.927338                 43.990574             1448.492500      1.335989     upper  43.991231  32.927338        96.151997      True             0.000000  minicircle.spm
5                432.420548  215.785933    2.875581   23.466926    15.068693      15.781069    0.806992    3.245209       2.193733     2.045240  1238.811017  605.704536          1238.253351                20.607085                 45.698864              941.720385      2.217629     upper  46.016578  20.607085        93.905352     False            22.980430  minicircle.spm
6                194.525710  231.200777    9.008959   23.627316    17.816803      18.014840    0.807410    2.893747       2.056074     1.935221  1291.656345  667.446377          1559.408536                29.666377                 45.282153             1343.357407      1.526380     upper  45.966170  29.666377        98.564277      True             0.000000  minicircle.spm
7                389.738847  274.400104    2.076301   25.556395    16.662449      18.408240    0.806428    3.043659       2.228860     2.058627  1440.338234  699.659511          1770.258141                22.716236                 49.213645             1117.948783      2.166452     upper  49.402763  22.716236        42.321695      True             0.000000  minicircle.spm
8                430.344334  295.941830    2.896435   22.126524    15.426546      16.421498    0.806669    4.426402       2.229832     2.113012  1382.992206  654.512315          1420.794444                26.255766                 42.786388             1123.389372      1.629600     upper  43.023656  26.255766        56.701634      True             0.000000  minicircle.spm
9                122.677158  308.875465    8.653363   27.439540    17.463583      16.648976    0.811939    3.034289       2.237907     2.050882  1393.878598  679.648321          1706.319951                25.553732                 52.364206             1338.100905      2.049180     upper  52.562005  25.553732        76.021062      True             0.000000  minicircle.spm
10                39.395704  381.913890    1.852146   26.940401    16.797841      17.346986    0.807436    2.851219       2.135492     1.982902  1276.542183  643.774604          1197.010778                19.381958                 53.361547             1034.251245      2.753156     upper  53.480238  19.381958       100.815523     False            28.995068  minicircle.spm
11               390.261574  377.074286    9.219255   26.733505    17.572591      16.718590    0.810151    2.989053       2.187163     2.017077  1314.294676  651.583848          1449.591034                27.759526                 47.482741             1318.098383      1.710503     upper  47.793347  27.759526        83.072558      True             0.000000  minicircle.spm
12               261.282576  374.871963    5.851948   25.596843    17.634806      18.608781    0.807326    2.863795       2.091227     1.947446  1320.726852  678.184088          1452.519500                25.341383                 49.572984             1256.247964      1.956207     upper  50.011545  25.341383       100.508933      True             0.000000  minicircle.spm
13               304.709441  391.099345   11.262835   26.558465    18.533693      18.899549    0.808821    3.101631       2.051124     1.933581  1276.877612  660.369249          1665.565455                35.564081                 42.264035             1503.081539      1.188391     upper  48.968624  33.471983        98.690295      True             0.000000  minicircle.spm
14               455.525239  428.028596    4.449485   26.611907    16.950644      17.544803    0.806675    4.008602       2.218208     2.077857  1345.784592  647.679226          1319.762342                21.630208                 51.381705             1111.396954      2.375461     upper  51.520977  21.630208       113.771377     False            16.309580  minicircle.spm
15               220.372355  418.201868    3.999989   27.385160    16.822496      16.318950    0.806871    3.196942       2.120307     1.984553  1352.672668  681.600632          1530.123869                25.443365                 51.342545             1306.327097      2.017915     upper  51.347797  25.443365        85.061755     False            27.703832  minicircle.spm
16               375.557982  428.615292    8.190056   25.036625    18.767056      19.279584    0.806910    3.093909       2.132358     1.982239  1403.823220  708.200872          1506.940174                30.628182                 45.448270             1391.997854      1.483871     upper  48.164652  30.628182       106.297141      True             0.000000  minicircle.spm
17               460.798356  458.626236    8.151931   27.857119    16.935121      16.569920    0.807411    3.160642       2.182733     2.019950  1272.787692  630.108426          1593.085904                21.720928                 52.398941             1138.153641      2.412371     upper  53.024260  21.720928        70.007748      True             0.000000  minicircle.spm
18               419.708892  456.361997   10.464533   24.875794    18.165731      18.788324    0.807342    3.053727       2.151053     2.001470  1362.249279  680.624477          1603.335537                34.580205                 42.978255             1486.196868      1.242857     upper  45.966170  33.555968       101.546870      True             0.000000  minicircle.spm
19                90.846917  468.962832    4.405668   24.980790    16.140796      16.473390    0.808337    2.940226       2.127244     1.983076  1291.172544  651.095771          1600.163031                21.103778                 48.632833             1026.336499      2.304461     upper  48.656154  21.103778       102.506168     False            14.427930  minicircle.spm
20                19.800613  501.342661    2.017733   25.451835    16.221300      17.120073    0.807429    2.996698       2.153397     2.004426  1199.415140  598.383369          1342.213920                20.183903                 49.933420             1007.851308      2.473923     upper  50.013985  20.183903        94.323563     False            19.291431  minicircle.spm
